packages/gatsby/src/db/index.js

```diff
@@ -0,0 +1,276 @@
+const _ = require(`lodash`)
+const fs = require(`fs-extra`)
+const path = require(`path`)
+const invariant = require(`invariant`)
+const loki = require(`lokijs`)
+const lokiFsStructuredAdapter = require(`lokijs/src/loki-fs-structured-adapter`)
+
+/////////////////////////////////////////////////////////////////////
+// DB Initialization
+/////////////////////////////////////////////////////////////////////
+
+// Must be set using `start()`
+let db
+
+function startDb(saveFile) {
+  return new Promise((resolve, reject) => {
+    const dbOptions = {
+      autoload: true,
+      autoloadCallback: err => {
+        if (err) {
+          reject(err)
+        } else {
+          resolve()
+        }
+      },
+      autosave: true,
+      autosaveInterval: 1000,
+    }
+    db = new loki(saveFile, dbOptions)
+  })
+}
+
+/**
+ * Starts a loki database. If the file already exists, it will be
+ * loaded as the database state. If not, a new database will be
+ * created.
+ *
+ * @param {string} saveFile on disk file that the database will be
+ * saved and loaded from.
+ * @returns {Promise} promise that is resolved once the database and
+ * (optionally) the existing state has been loaded
+ */
+async function start({ saveFile }) {
+  if (!_.isString(saveFile)) {
+    throw new Error(`saveFile must be a path`)
+  }
+  const saveDir = path.dirname(saveFile)
+  await fs.ensureDir(saveDir)
+  await startDb(saveFile)
+}
+
+/**
+ * Returns a reference to the database. If undefined, the db has not been
+ * initalized yet. Call `start()`
+ *
+ * @returns {Object} database, or undefined
+ */
+function getDb() {
+  return db
+}
+
+/**
+ * Deletes all collections.
+ */
+function deleteAllCollections() {
+  if (db) {
+    _.forEach(db.listCollections(), collInfo => {
```

Comment:
Might be worth benchmarking _.forEach vs. native forEach (in general, not just here!). I tend to prefer to defer to native array methods, but I may be unique in that regard.

I think that it's a safer bet to presume they'll be optimized very efficiently, if they're not already!

Reply:
Cool, I'm actually a relative newcomer to js so any tips like this are greatly appreciated. I may have gone a bit lodash mad in this PR as I come from a functional background. Is this the native for each you mean? https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Statements/for...of

Reply:
Silly question, is db.listCollections() an _object_ or an _array_? I presumed array, but on second glance, it could actually be an object?

If it's the latter, then this _.forEach may actually be preferable!

I'm about to board a plane, but I'll add some more feedback on this tomorrow. Nice work!!

Reply:
Sorry, forgot to reply. It's an array. I've been using `for (const foo of someArray)` and I'm definitely preferring it